File: weektwo/revise/sorting/Merge.java
package revise.sorting;

import java.util.Arrays;

public class Merge {
    public static void main(String[] args) {
        int[] arr = { 2, 3, 1, 5, 4 };
        sort(arr);
        System.out.println(Arrays.toString(arr));
    }

    private static void sort(int[] arr) {
        int totalLength = arr.length;
        int mid = totalLength / 2;

        if (totalLength < 2) {
            return;
        }

        int left[] = new int[mid];
        int right[] = new int[totalLength - mid];

        for (int i = 0; i < mid; i++) {
            left[i] = arr[i];
        }

        for (int i = mid; i < totalLength; i++) {
            right[i - mid] = arr[i];
        }

        sort(left);
        sort(right);

        merge(arr, left, right);
    }

    private static void merge(int[] arr, int[] left, int[] right) {
        int leftLength = left.length;
        int rightLength = right.length;
        int i = 0, j = 0, k = 0;
        while (leftLength > i && rightLength > j) {
            if (left[i] > right[j]) {
                arr[k] = right[j];
                j++;
            } else {
                arr[k] = left[i];
                i++;
            }
            k++;
        }

        while (leftLength > i) {
            arr[k] = left[i];
            i++;
            k++;
        }

        while (rightLength > j) {
            arr[k] = right[j];
            j++;
            k++;
        }
    }
}

File: dsaJavascript/general/merge.js
function merge(arr) {
  let length = arr.length;

  if (length < 2) {
    return arr;
  }

  let mid = Math.floor(length / 2);
  let left = arr.slice(0, mid);
  let right = arr.slice(mid);

  return sort(merge(left), merge(right));
}

function sort(left, right) {
  let newArray = [];
  let i = 0;
  let j = 0;

  while (left.length > i && right.length > j) {
    if (left[i] > right[j]) {
      newArray.push(right[j]);
      j++;
    } else {
      newArray.push(left[i]);
      i++;
    }
  }

  while(left.length > i){
    newArray.push(left[i]);
    i++
  }

  while(right.length > j){
    newArray.push(right[j]);
    j++
  }

  return newArray
}

let arr = [12, 3, 1, 923, 2];
arr=merge(arr);
console.log(arr);
